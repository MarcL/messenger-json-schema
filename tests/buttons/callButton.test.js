const ajv = require('../ajv');
const callButton = require('../../messenger/buttons/callButton');

describe('Buttons: call button', () => {
    let validate;
    
    beforeEach(() => {
        validate = ajv.compile(callButton);
    });

    test('Valid message with phone number', () => {
        const givenMessage = {
            type: 'phone_number',
            title: 'Call mum',
            payload: '+12345'
        };

        validate(givenMessage);
        
        expect(validate.errors).toBeNull();
    });

    describe('Invalid message', () => {
        const invalidScenarios = [
            {
                testMessage: 'Message type is invalid',
                givenMessage: {
                    type: 'not_phone_number',
                    title: 'Call mum',
                    payload: '+12345'
                }
            },
            {
                testMessage: 'Message missing type property',
                givenMessage: {
                    title: 'Call mum',
                    payload: '+12345'
                }
            },
            {
                testMessage: 'Message missing title property',
                givenMessage: {
                    type: 'not_phone_number',
                    payload: '+12345'
                }
            },
            {
                testMessage: 'Message missing payload property',
                givenMessage: {
                    type: 'phone_number',
                    title: 'Call mum',
                }
            },
            {
                testMessage: 'Message has invalid payload phone number',
                givenMessage: {
                    type: 'phone_number',
                    title: 'Call mum',
                    payload: '12356',
                }
            },
            {
                testMessage: 'Message contains additional properties',
                givenMessage: {
                    type: 'phone_number',
                    title: 'Call mum',
                    payload: '12356',
                    invalid: 'Invalid property'
                }
            }
        ];

        invalidScenarios.forEach(scenario => {
            const {testMessage, givenMessage} = scenario;

            test(testMessage, () => {        
                validate(givenMessage);

                expect(validate.errors).toMatchSnapshot();
            });
        });

        test('Message title is greater than 20 characters', () => {
            const title = new Array(21).fill('a').join('');
            const givenMessage = {
                type: 'phone_number',
                title,
                payload: '12356'
            };

            validate(givenMessage);

            expect(validate.errors).toMatchSnapshot();
        });
    });
});
