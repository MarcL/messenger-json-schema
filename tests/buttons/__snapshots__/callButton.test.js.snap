// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons: call button Invalid message Message contains additional properties 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "invalid",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "/payload",
    "keyword": "pattern",
    "message": "should match pattern \\"^\\\\+[0-9]+\\"",
    "params": Object {
      "pattern": "^\\\\+[0-9]+",
    },
    "schemaPath": "#/properties/payload/pattern",
  },
]
`;

exports[`Buttons: call button Invalid message Message has invalid payload phone number 1`] = `
Array [
  Object {
    "dataPath": "/payload",
    "keyword": "pattern",
    "message": "should match pattern \\"^\\\\+[0-9]+\\"",
    "params": Object {
      "pattern": "^\\\\+[0-9]+",
    },
    "schemaPath": "#/properties/payload/pattern",
  },
]
`;

exports[`Buttons: call button Invalid message Message missing payload property 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'payload'",
    "params": Object {
      "missingProperty": "payload",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`Buttons: call button Invalid message Message missing title property 1`] = `
Array [
  Object {
    "dataPath": "/type",
    "keyword": "enum",
    "message": "should be equal to one of the allowed values",
    "params": Object {
      "allowedValues": Array [
        "phone_number",
      ],
    },
    "schemaPath": "#/properties/type/enum",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'title'",
    "params": Object {
      "missingProperty": "title",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`Buttons: call button Invalid message Message missing type property 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'type'",
    "params": Object {
      "missingProperty": "type",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`Buttons: call button Invalid message Message title is greater than 20 characters 1`] = `
Array [
  Object {
    "dataPath": "/title",
    "keyword": "maxLength",
    "message": "should NOT be longer than 20 characters",
    "params": Object {
      "limit": 20,
    },
    "schemaPath": "#/properties/title/maxLength",
  },
  Object {
    "dataPath": "/payload",
    "keyword": "pattern",
    "message": "should match pattern \\"^\\\\+[0-9]+\\"",
    "params": Object {
      "pattern": "^\\\\+[0-9]+",
    },
    "schemaPath": "#/properties/payload/pattern",
  },
]
`;

exports[`Buttons: call button Invalid message Message type is invalid 1`] = `
Array [
  Object {
    "dataPath": "/type",
    "keyword": "enum",
    "message": "should be equal to one of the allowed values",
    "params": Object {
      "allowedValues": Array [
        "phone_number",
      ],
    },
    "schemaPath": "#/properties/type/enum",
  },
]
`;
